### PR TITLE
libnet/d/bridge: Allow IPv6 ICC from any IP address

### PR DESCRIPTION
**Pull request by akerouanton**
Fixes:

- #45460

**- What I did**

IPv6 ipt rules are exactly the same as IPv4 rules, although both protocol don't use the same networking model. This has bad consequences, for instance: 1. the current v6 rules disallow Neighbor Solication/Advertisement ; 2. multicast addresses can't be used ; 3. link-local addresses are blocked too.

To solve this, this commit changes the following rules:

```
-A DOCKER-ISOLATION-STAGE-1 ! -s fdf1:a844:380c:b247::/64 -o br-21502e5b2c6c -j DROP
-A DOCKER-ISOLATION-STAGE-1 ! -d fdf1:a844:380c:b247::/64 -i br-21502e5b2c6c -j DROP
```

into:

```
-A DOCKER-ISOLATION-STAGE-1 ! -s fdf1:a844:380c:b247::/64 ! -i br-21502e5b2c6c   -o br-21502e5b2c6c -j DROP
-A DOCKER-ISOLATION-STAGE-1 ! -d fdf1:a844:380c:b247::/64   -i br-21502e5b2c6c ! -o br-21502e5b2c6c -j DROP
```

These rules only limit the traffic ingressing/egressing the bridge, but not traffic between veth on the same bridge.

Note that, the kernel takes care of dropping invalid IPv6 packets, eg. loopback spoofing, thus these rules don't need to be more specific.

**- How to verify it**

```
$ docker network create  --subnet 'fdf1:a844:380c:b247::/64' --ipv6 --internal testnet
$ docker run --rm -d --network testnet --name test1 nicolaka/netshoot /bin/sleep infinity
# Both ping should work
$ docker run --rm -t --network testnet --name test2 nicolaka/netshoot ping -c1 -6 test1
$ docker run --rm -t --network testnet --name test2 nicolaka/netshoot ping -c1 -6 fe80::42:acff:fe12:2%eth0
```

To check loopback spoofing:

```console
$ docker run --rm -d --network testnet --name test1 nicolaka/netshoot nc -6lk 1234
$ docker run --rm -t --network testnet --name test2 akerouanton/scapy-scripts:latest ip-spoofer.py \
    --ip-family ipv6 \
	--src-ip fdf1:a844:380c:b247::3 --dst-ip ::1 \
	--dst-port 1234 \
	--src-mac-addr "02:42:ac:12:00:03" --dst-mac-addr "02:42:ac:12:00:02" \
	--iface eth0
Received 4 packets, got 0 answers, remaining 1 packets
```

**- Description for the changelog**

Bridge networks now allow packets with link-local, multicast or unspecified IP address to be sent to other hosts on the same network. This fixes a bug where IPv6 Neighbor Solicitation wouldn't work within internal networks.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: libnetwork/drivers/bridge/setup_ip_tables.go
Comment (corhere) [inline libnetwork/drivers/bridge/setup_ip_tables.go:426]: Shouldn't these be `iptables.IPv6`? (Or `version`?) Why are there separate sets of `programChainRule` calls for v4 vs. v6, but only a single shared `setIcc` call?

Comment (akerouanton) [reply libnetwork/drivers/bridge/setup_ip_tables.go]: > Shouldn't these be iptables.IPv6?

Doh :facepalm: Should be fixed now.

> Why are there separate sets of programChainRule calls for v4 vs. v6, but only a single shared setIcc call?

Not sure I get your question. `setIcc` only adds a rule that look like this:

```
-A FORWARD -i <bridge-iface> -o <bridge-iface> -j ACCEPT
```

Hence there's no need to program iptables in different ways.

**Review by corhere — APPROVED** [review]

**Review by neersighted — APPROVED** [review]
LGTM; not sure we want to backport as this is borderline fixing a bug/a new behavior. Still, happy to take a backport if we think this is universally undesirable.

**Comment by akerouanton** [comment]
I'll put this on hold as I'd like to make some more tests here before we merge it.

**Comment by akerouanton** [comment]
I cherry-picked this on top of #45850 and unskipped the following tests. They're now passing :tada:

* `TestBridgeICC/IPv6_ULA_on_non-internal_network`
* `TestBridgeICC/IPv6_link-local_address_on_internal_network`
* `TestBridgeIPv6SLAACLLAddress/IPv6_internal_network`

**Review by corhere — APPROVED** [review]

**Pull request merged**
